ddl/column_type_change_test.go

```diff
@@ -461,6 +469,38 @@ func (s *testColumnTypeChangeSuite) TestColumnTypeChangeBetweenVarcharAndNonVarc
 	tk.MustGetErrCode("alter table t change column a a varchar(10);", mysql.ErrUnsupportedDDLOperation)
 	tk.MustGetErrCode("alter table t change column b b char(10);", mysql.ErrUnsupportedDDLOperation)
 	tk.MustExec("admin check table t;")
+
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(a char(10), b varchar(10));")
+	tk.MustExec("insert into t values ('aaa    ', 'bbb   ');")
+	tk.MustExec("alter table t change column a a varchar(10);")
+	tk.MustExec("alter table t change column b b char(10);")
+
+	tk.MustExec("alter table t add index idx_a(a);")
+	tk.MustExec("alter table t add index idx_b(b);")
+	tk.MustExec("alter table t change column a a char(10);")
+	tk.MustExec("alter table t change column b b varchar(10);")
+	tk.MustExec("admin check table t;")
+
+	// https://github.com/pingcap/tidb/issues/23624
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(b varchar(10));")
+	tk.MustExec("insert into t values ('aaa   ');")
+	tk.MustExec("alter table t change column b b char(10);")
+	tk.MustExec("alter table t add index idx(b);")
+	tk.MustExec("alter table t change column b b varchar(10);")
+	tk.MustQuery("select b from t use index(idx);").Check(testkit.Rows("aaa"))
+	tk.MustQuery("select b from t ignore index(idx);").Check(testkit.Rows("aaa"))
+	tk.MustExec("admin check table t;")
+
+	// https://github.com/pingcap/tidb/pull/23688#issuecomment-810166597
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a varchar(10));")
+	tk.MustExec("insert into t values ('aaa   ');")
+	tk.MustQuery("select a from t;").Check(testkit.Rows("aaa   "))
+	tk.MustExec("alter table t modify column a char(10);")
+	tk.MustQuery("select a from t;").Check(testkit.Rows("aaa"))
```

Comment:
It seems that this case duplicated with line488-490. Could we add `select` statement after line490 directly?

Reply:
The first case shows consistency between index data and row data. The second case shows that it should get a different result in different types with the same data.

Let's keep it separate to clarify the problems.

Reply:
OK~